### PR TITLE
Fixes #26953 - extract base and build-base container images

### PR DESCRIPTION
**Pull request by iNecas**
This patch introduces `Dockerfile.builder-base` file that can produce
the following images:

* *quay.io/foreman/foreman:base* - the base os used in all of the other
    container images
* *quay.io/foreman/foreman:builder-base* - the image with all basic ruby
    and node dependencies to streamline the build process

To test locally, run this before building the main image:

```
docker build --target base -t quay.io/foreman/foreman-base \
  -f Dockerfile.builder-base .
docker build -t quay.io/foreman/foreman-builder-base \
  -f Dockerfile.builder-base .
```

### CODE REVIEW

**Comment by iNecas** [comment]
Btw, while touching the file structure, what do you think about moving the containers-related files to `containers` subdirectory, that could also include the README, that would link the `developer_doc` or vice-verse. I guess some helper scripts for actually building the stuff might be useful as well (not necessary part of this PR though)

File: Dockerfile [hunk 1]
```diff
@@ -1,71 +1,35 @@
...
+RUN \
+  bundle install && \
+  npm install --no-optional
+
 RUN \
   make -C locale all-mo && \
   bundle exec rake assets:clean assets:precompile db:migrate &&  \
-  bundle exec rake db:seed apipie:cache:index && rm tmp/bootstrap-db.sql
+  bundle exec rake db:seed apipie:cache:index && rm tmp/bootstrap-db.sqlite3
 RUN ./node_modules/webpack/bin/webpack.js --config config/webpack.config.js && npm run analyze && rm -rf public/webpack/stats.json
```
Comment (ohadlevy) [inline Dockerfile]: if you are running it again, I assume you would need to remove the old assets first? otherwise in the image you will have two versions? (one for each time you compiled assets?) probably rm -rf public/webpack/* will work.

File: Dockerfile.builder-base [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    -c "Foreman Application User" foreman && \
+    chown -R 1001:1001 $HOME
+
+# Add a script to be executed every time the container starts.
+COPY entrypoint.sh /usr/bin/
+RUN chmod +x /usr/bin/entrypoint.sh
+ENTRYPOINT ["entrypoint.sh"]
+
+# Temp container that download gems/npms and compile assets etc
+FROM base as builder-base
```
Comment (ohadlevy) [inline Dockerfile.builder-base]: no need for "as" here?

File: Dockerfile (hunk 1 above)
Comment (iNecas) [reply Dockerfile]: Note that we are not running in the the builder-base image: the builder-base is just to pre-load the dependencies

File: Dockerfile.builder-base (hunk 2 above)
Comment (iNecas) [reply Dockerfile.builder-base]: Updated

**Comment by iNecas** [comment]
[skip ci]
[skip build]

**Comment by ohadlevy** [comment]
> Btw, while touching the file structure, what do you think about moving the containers-related files to `containers` subdirectory, that could also include the README, that would link the `developer_doc` or vice-verse. I guess some helper scripts for actually building the stuff might be useful as well (not necessary part of this PR though)

I don't mind moving into a folder,  I could see the following in the subdirectory:
- /containers/Dockerfile
- /containers/.dockerignore 
- /containers/entrypoint.sh
- /containers/docker-compose.yml (even if feels more natural to be in the root) \
- /containers/kubernetes/... yaml examples (here are some https://github.com/ohadlevy/foreman/tree/dockerfile/kube)

I'm uncertain how we would handle triggering builds of the non production image(currently every time we merge into the repo we build a new image based on the branch name), perhaps, as part of this pr, we need to clearly outline which user stories we are trying to solve? e.g.
- have a faster image build time (to avoid installing ruby/node/python all the time) 
- have a image that can be used for development
- speed up plugin image build time by reusing a foreman published build image

then we should go case by case and figure out how we publish it, how to use it etc? (perhaps, discourse is the better place for it..)
WDYT?

**Comment by ohadlevy** [comment]
@iNecas I think that you actually need to break it down to 3 dockerfiles - as now we are not using compose for base/builder (and introduced a new temp container).

have a look at https://github.com/ohadlevy/foreman/commit/2a07924576ea1c6af828240a5ea2f66aaab822ff this enable 3 "published" containers - foreman-base, foreman-builder and foreman.
I've setup my quay account to build all 3 containers upon a repository push, not efficient for quay, for works for now. it does have a limitation that foreman builds are usually using a build-1 image of the builder and builder uses a base - 2 image of base.

I didn't try the resulting images (assuming they work correctly etc) YMMV.

File: containers/README.asciidoc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+and then, use tar to combine all needed directories and follow symlinks, e.g.
+
+[source, bash]
+....
+tar -chf - --exclude-from=.dockerignore . | docker build -
+....
+
+*Buildah note*: In case you wanted to build the container image with Buildah, be aware of https://github.com/containers/buildah/issues/1578[the following issue] that currently prevents using it.
+
+==== Building base images
```
Comment (iNecas) [inline containers/README.asciidoc]: The rest of the docs is unchnaged: I've just added this section on how the layers are composed and how to build them locally

File: containers/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,48 @@
+FROM quay.io/foreman/foreman-builder-base as builder
```
Comment (ohadlevy) [inline containers/Dockerfile]: is it builder or base? imho the naming are confusing.

File: containers/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+USER 1001
+WORKDIR $HOME
+COPY --chown=1001:1001 . ${HOME}/
+COPY --from=builder /usr/bin/entrypoint.sh /usr/bin/entrypoint.sh
+COPY --from=builder --chown=1001:1001 ${HOME}/.bundle/config ${HOME}/.bundle/config
+COPY --from=builder --chown=1001:1001 ${HOME}/Gemfile.lock ${HOME}/Gemfile.lock
+COPY --from=builder --chown=1001:1001 ${HOME}/vendor/ruby ${HOME}/vendor/ruby
+COPY --from=builder --chown=1001:1001 ${HOME}/public ${HOME}/public
+RUN echo gem '"rdoc"' > bundler.d/container.rb && echo gem '"tzinfo-data"' >> bundler.d/container.rb
```
Comment (ohadlevy) [inline containers/Dockerfile]: I would prefer to copy it instead of recreate 
```
COPY --from=builder --chown=1001:1001 ${HOME}/bundler.d/container.rb ${HOME}/bundler.d/container.rb
```

File: containers/Dockerfile.builder-base [hunk 6]
```diff
@@ -2,8 +2,6 @@
 FROM registry.fedoraproject.org/fedora-minimal:30 as base
 ARG RUBY_VERSION="2.6"
 ARG NODEJS_VERSION="11"
-ENV FOREMAN_FQDN=foreman.example.com
-ENV FOREMAN_DOMAIN=example.com
```
Comment (ohadlevy) [inline containers/Dockerfile.builder-base:6]: why? this will fail the production image (its needed to build and to run) if you don't include it here.

File: containers/README.asciidoc (hunk 3 above)
Comment (ohadlevy) [reply containers/README.asciidoc]: can you use git move instead? its hard to follow the changes.

**Review by ohadlevy — CHANGES_REQUESTED** [review]
left some inline comments, a few questions.

- did you test this with quay? which additional hooks / steps are required? (at least to add the two new images as new repositories).
- I'm in mixed feelings about the docker-compose (its useful to be in the parent directory) and regardless you would at least need to change the build context at https://github.com/theforeman/foreman/blob/136a8c19124beb653dc0499f33ab3e0670d3a8cb/containers/docker-compose.yml#L26 to ```..``` as I assume its currently broken.

File: containers/Dockerfile (hunk 5 above)
Comment (iNecas) [reply containers/Dockerfile]: Just copied over from what there already was before, but np, updated

File: containers/README.asciidoc (hunk 3 above)
Comment (iNecas) [reply containers/README.asciidoc]: Git mv by itself doesn't help much for the overall diff. Anyway, I've moved the move to separate commit, so you can see the actual changes here https://github.com/theforeman/foreman/pull/6825/commits/f743423084601f2f78570201e9358d7447d4ef03

File: containers/Dockerfile (hunk 4 above)
Comment (iNecas) [reply containers/Dockerfile]: It's `builder-base`: the base for the `builder`. the `builder` itself is the container being built. I'm hesitant calling the `foreman-builder-base` as `builder` because, technically, the `builder` image is the product of `builder` phase in `Dockerfile` (same as base is not the fedora-minimal but fedora-miniman + additional changes we need)

**Comment by iNecas** [comment]
@ohadlevy I've tested the building locally, still need to try with quay: can I create the repos in the foreman org already, or should I test it in my namespace?

I've updated the docker-compose as well: should be working now

**Comment by iNecas** [comment]
Here is the discourse thread asking for more opinions on the proposal + build strategy https://community.theforeman.org/t/foreman-dockerfile-proposal-publish-base-and-builder-base-images/14206

File: containers/Dockerfile (hunk 4 above)
Comment (ohadlevy) [reply containers/Dockerfile]: imho foreman-builder makes more sense :)

File: containers/Dockerfile.builder-base (hunk 6 above)
Comment (iNecas) [reply containers/Dockerfile.builder-base:6]: Just wrong conflict resolution (had it in different image). Fixed

File: containers/Dockerfile (hunk 4 above)
Comment (iNecas) [reply containers/Dockerfile]: Leaving some time for anyone else to express their opinion :)

Comment (ohadlevy) [reply containers/Dockerfile]: and to answer to more than oneliner, from a consumer POV:

foreman image:
temp builder container: FROM quay.io/foreman/foreman-builder
production container: FROM quay.io/foreman/foreman-base

just seems clear one is image is a base image, the other is the builder image, I understand the nuance that you build later on on top using another layered image, but think about the guy that browse the published images and try to figure out what is what.

File: containers/README.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    the base OS with minimal runtime (ruby, node, python) dependencies.
+quay.io/theforeman/foreman-builder-base::
+    the base image for the builder image, with pre-installed devel/build dependencies
+builder::
+    https://docs.docker.com/develop/develop-images/multistage-build/[a staged image]
+    to build the assets and other built-time artifacts to be copied over to the final image
+
+
+....
+Dockerfile.builder-base[quay.io/theforeman/foreman-base]
```
Comment (ohadlevy) [inline containers/README.md]: should be just base right? (see how its confusing :-)

File: containers/Dockerfile (hunk 4 above)
Comment (ohadlevy) [reply containers/Dockerfile]: aha! I understand my cnofusion! there should be 3 published images -> 3 dockerfles! please have another look at https://github.com/ohadlevy/foreman/commit/2a07924576ea1c6af828240a5ea2f66aaab822ff

Comment (iNecas) [reply containers/Dockerfile]: I'm ok with 3 Dockerfiles

**Comment by ohadlevy** [comment]
@iNecas coming back to this, whats the status of this PR? thanks.

**Comment by iNecas** [comment]
@ohadlevy updated: 3 Dockefiles and renamed `foreman-builder-base` to `foreman-builder`

**Comment by ares** [comment]
I've checked with @iNecas and there's very little chance he could get to it in near feature. Therefore we're closing this PR. Thanks @iNecas for your effort on this, hopefully one day someone will use it as a base.
